### PR TITLE
Add team and projects APIs

### PR DESCRIPTION
**Pull request by andrerpbts**
This PR implements the grape gem in order to supply Team and Projects APIs

You can manage APITokens in /manage or in Team edit form.

If a api_token is linked with a team, this use will be restrict by this access. Otherwise, this will have full access.

The API explorer is located in `/api/v1/explore`. 

Closes https://github.com/Codeminer42/cm42-central/issues/73

captura de tela 2016-12-08 as 23 48 03

captura de tela 2016-12-08 as 23 48 24

captura de tela 2016-12-08 as 23 48 32

### CODE REVIEW

File: app/api/v1/projects.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    before do
+      authenticate!
+      authorize_project!
+    end
+
+    desc 'Return all projects', { tags: ['project'] }
+    paginate
+    get '/' do
+      projects = Project.all
```
Comment (daniloisr) [inline app/api/v1/projects.rb]: I'm not familiar with Grape, but seems that would be a great idea to use Pundit on API endpoints too.

Comment (andrerpbts) [inline app/api/v1/projects.rb]: Maybe it's possible, but I don't know right how to do it.

File: app/api/v1/projects.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
+class V1::Projects < Grape::API
+  resource :projects do
+    helpers do
+      def authorize_project!
+        return unless current_team
+        @allowed_projects = current_team.projects.pluck(:slug)
+
+        return unless params[:slug]
+        return if @allowed_projects.include?(params[:slug])
```
Comment (adbatista) [inline app/api/v1/projects.rb]: return if !params[:slug] || @allowed_projects.include?(params[:slug])`

File: app/controllers/api_tokens_controller.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+class ApiTokensController < ApplicationController
+  before_action :set_team
+
+  def create
+    @api_token = @team.api_tokens.create
+    redirect_to edit_team_path(@team)
```
Comment (adbatista) [inline app/controllers/api_tokens_controller.rb]: redirect in API?

File: spec/api/v1/health_spec.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+require 'rails_helper'
+
+RSpec.describe V1::Health do
+  describe '#GET check' do
+    subject { get '/api/v1/health' }
+
+    it 'returns ok' do
+      subject
```
Comment (adbatista) [inline spec/api/v1/health_spec.rb]: put the get direct here

File: spec/api/v1/projects_spec.rb [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+  describe '#GET /api/v1/projects' do
+    let(:team) { create :team }
+    subject { get '/api/v1/projects', per_page: 2, api_key: api_token.token }
+
+    before do
+      create_list :project, 4, teams: [team]
+    end
+
+    it 'returns 2 projects' do
+      subject
```
Comment (adbatista) [inline spec/api/v1/projects_spec.rb]: the same, here, it makes the test easier to read, or if you add a name to the subject, I'm ok too

File: app/controllers/api_tokens_controller.rb (hunk 3 above)
Comment (andrerpbts) [inline app/controllers/api_tokens_controller.rb]: It isn't the API, but token controller, to manage API tokens.

**Comment by akitaonrails** [comment]
1) Fix the N+1 such as:

```
GET /api/v1/projects/hiroshima?api_key=4549e330b63e763834d01235b2aceea2
N+1 Query detected
  ApiToken => [:team]
  Add to your finder: :includes => [:team]
N+1 Query method call stack
  /vagrant/codeminer42/central/central/app/api/api.rb:17:in `current_team'
  /vagrant/codeminer42/central/central/app/api/v1/projects.rb:5:in `authorize_project!'
  /vagrant/codeminer42/central/central/app/api/v1/projects.rb:17:in `block (2 levels) in <class:Projects>'

/vagrant/codeminer42/central/central/app/api/api.rb:17:in `current_team'
/vagrant/codeminer42/central/central/app/api/v1/projects.rb:5:in `authorize_project!'
/vagrant/codeminer42/central/central/app/api/v1/projects.rb:17:in `block (2 levels) in <class:Projects>'
```

2) make the /projects/:slug endpoint also return how many stories pages there are and how many stories per page there are, otherwise how can we know how many pages to fetch from /projects/:slug/stories ?

3) add the option to pass the auth token through the header, for example, using Rails' built-in [`authenticate_or_request_with_http_token`](http://api.rubyonrails.org/classes/ActionController/HttpAuthentication/Token.html) so that you can authenticate like this:

```ruby
  def authenticate
    authenticate_or_request_with_http_token do |token, options|
      @api_key ||= ApiToken.find_by(token: token)
    end
  end
```

And then you can test it like this: `curl -H "Authorization: Token token=WwEsPpOCYMsyPsmmvKBqQDOaEJ4t" http://localhost:3000/api/v1/projects`

4) Add a `/projects/:slug/analysis` endpoint to concentrate the `IterationService` public methods such as `velocity`, `volatility`, `backlog_iterations`, `current_iteration_number`, `backlog`, including the attributes in the initializer (project, since and current_time).

cc @andrerpbts

**Comment by rodrigovirgilio** [comment]
1. Done.

2. So I understood the gem itself already does this, see the headers project hiroshima:
{
  "x-runtime": "0.205583",
  **"total": "44",
  "per-page": "25",**
  "etag": "W/\"5fdbefcbb892ea21a8e7fccfdc7e09bd\"",
  "vary": "Origin",
  "content-type": "application/json",
  "cache-control": "max-age=0, private, must-revalidate",
  "link": "<http://localhost:3000/api/v1/projects/hiroshima/stories?api_key=621926363b8f5d168b72808a660f04da&page=2&state=backlog>; rel=\"last\", <http://localhost:3000/api/v1/projects/hiroshima/stories?api_key=621926363b8f5d168b72808a660f04da&page=2&state=backlog>; rel=\"next\"",
  "content-length": "13030",
  "x-request-id": "f6aa2d54-68d7-4049-93f4-bb17924c29cf"
}

3. https://github.com/Codeminer42/cm42-central/pull/77/files#diff-8812cec49240789009b71b1ff58b4451R6

4. https://github.com/Codeminer42/cm42-central/pull/77/files#diff-f26cf1da722d7f1d3212c9363eacaa7cR40

https://github.com/Codeminer42/cm42-central/pull/77/files#diff-3fc6babe2f87b2b949d120e9698c8978

cc @akitaonrails

**Comment by rodrigovirgilio** [comment]
Ok, done squash of the commits.

**Pull request merged**